### ISSUE TITLE
FI trtllm_fp4_block_scale_moe interface core dump and output mismatch

### ISSUE THREAD

**Issue by rosong11** [flashinfer-ai/flashinfer#2373]
With nvfp4 precision in multi-GPU tests, when routing_method_type is set to DeepSeekV3, the fp4_block_scale_moe_runner interface core dumps unless routing_bias is provided but trt-llm doesn't need it. 
I found that this is because when `routing_logits` is None, we currently have to manually pack the `topk_weights` and `topk_ids` and pass them into FI’s fused_moe API, whereas the trtllm-gen kernel in TRT-LLM supports passing these two arguments directly. As a result, the two paths enter different kernels for routing-related computation. In the FI path, the kernel performs an unnecessary access to `routing_bias` and lacks a null check[here.](https://github.com/flashinfer-ai/flashinfer/blob/885fbf53d70256db2040379ccb4c881baa624fe3/csrc/trtllm_fused_moe_routing_deepseek.cu#L75C5-L75C101) 
**Could we add support on the FI side for passing topk_weights and topk_ids as parameters?**
After passing in `routing_bias` to fix the core dump, **the results from the two sides still don’t match**. For reference, here are the printed kernel_names from both sides.
[TRTLLM]
[TensorRT-LLM][INFO] [PID 238888] NumBatches 72, MaxNumCtasInBatchDim 1607, ShapeMNK 32768 768 2560, ValidShapeMNK 32768 768 2560, Kernel bmm_E2m1_E2m1E2m1_Fp32_t128x128x256u2_s6_et128x32_m256x128x64_cga2x1x1_16dp256b_rM_TN_transOut_schedP2x1x2x3_biasM_bN_tma_tmaOpt_clmp_swiGlu_dynBatch_sm100f
[TensorRT-LLM][INFO] [PID 238889] NumBatches 72, MaxNumCtasInBatchDim 1607, ShapeMNK 32768 768 2560, ValidShapeMNK 32768 768 2560, Kernel bmm_E2m1_E2m1E2m1_Fp32_t128x128x256u2_s6_et128x32_m256x128x64_cga2x1x1_16dp256b_rM_TN_transOut_schedP2x1x2x3_biasM_bN_tma_tmaOpt_clmp_swiGlu_dynBatch_sm100f
[TensorRT-LLM][INFO] [PID 238890] NumBatches 72, MaxNumCtasInBatchDim 1607, ShapeMNK 32768 768 2560, ValidShapeMNK 32768 768 2560, Kernel bmm_E2m1_E2m1E2m1_Fp32_t128x128x256_s6_et128x32_m256x128x64_cga2x1x1_16dp256b_rM_TN_transOut_schedP2x1x2x3_biasM_bN_tma_tmaOpt_clmp_swiGlu_dynBatch_sm100f
[TensorRT-LLM][INFO] [PID 238890] NumBatches 72, MaxNumCtasInBatchDim 1607, ShapeMNK 32768 2560 384, ValidShapeMNK 32768 2560 384, Kernel bmm_Bfloat16_E2m1E2m1_Fp32_t128x128x256_s6_et128x128_m256x128x64_cga2x1x1_16dp256b_rM_TN_transOut_schedP2x1x2x3_biasM_bN_tmaOpt_clmp_dynBatch_sm100f
[TensorRT-LLM][INFO] [PID 238890] NumBatches 72, MaxNumCtasInBatchDim 1607, ShapeMNK 32768 2560 384, ValidShapeMNK 32768 2560 384, Kernel bmm_Bfloat16_E2m1E2m1_Fp32_t128x128x256u2_s6_et128x128_m256x128x64_cga2x1x1_16dp256b_rM_TN_transOut_schedP2x1x2x3_biasM_bN_tmaOpt_clmp_dynBatch_sm100f
[TensorRT-LLM][INFO] [PID 238890] NumBatches 72, MaxNumCtasInBatchDim 1607, ShapeMNK 32768 768 2560, ValidShapeMNK 32768 768 2560, Kernel bmm_E2m1_E2m1E2m1_Fp32_t128x128x256u2_s6_et128x32_m256x128x64_cga2x1x1_16dp256b_rM_TN_transOut_schedP2x1x2x3_biasM_bN_tma_tmaOpt_clmp_swiGlu_dynBatch_sm100f
[FlashInfer]
=========FlshaInfer kernel name: bmm_E2m1_E2m1E2m1_Fp32_t128x128x256u2_s6_et128x32_m256x128x64_cga2x1x1_16dp256b_rM_TN_transOut_schedP2x1x2x3_bN_tma_tmaOpt_clmp_swiGlu_dynBatch_sm100f============
=========FlshaInfer kernel name: bmm_Bfloat16_E2m1E2m1_Fp32_t128x128x256u2_s6_et128x128_m256x128x64_cga2x1x1_16dp256b_rM_TN_transOut_schedP2x1x2x3_bN_tmaOpt_clmp_dynBatch_sm100f============
=========FlshaInfer kernel name: bmm_E2m1_E2m1E2m1_Fp32_t128x128x256u2_s6_et128x32_m256x128x64_cga2x1x1_16dp256b_rM_TN_transOut_schedP2x1x2x3_bN_tma_tmaOpt_clmp_swiGlu_dynBatch_sm100f============
=========FlshaInfer kernel name: bmm_Bfloat16_E2m1E2m1_Fp32_t128x128x256u2_s6_et128x128_m256x128x64_cga2x1x1_16dp256b_rM_TN_transOut_schedP2x1x2x3_bN_tmaOpt_clmp_dynBatch_sm100f============

**Comment by aleozlx** [comment]
let me see if i'm understanding it correctly

- FI lacks topk_weights and topk_ids params
- the mismatch after bypassing routing_bias issue is from the packing topk_weights and topk_ids usage and we end up getting different routing results. but we better first rectify api gap, adding topk_weights and topk_ids params first

**Comment by rosong11** [comment]
Thanks for your reply, It’s better to add these two parameters first.
It could be because we’re handling the packing of topk_weights and topk_ids differently, which leads to different routing results.
Another possibility is that we’re using an incorrect kernel version, the earlier issue where FI didn’t support bias was also specific to the NVFP4 precision path. 
I ran this unitest [here](https://github.com/NVIDIA/TensorRT-LLM/blob/9116dfbacdc87ed647d6cf703426601f1ecbf702/tests/integration/defs/accuracy/test_llm_api_pytorch.py#L1905)

**Comment by wenmingw** [comment]
@aleozlx  could you please help provide an ETA for this as well? Thanks.

**Comment by aleozlx** [comment]
re @wenmingw i'll put the ETA for this as end of month (1/31)

**Comment by aleozlx** [comment]
> Another possibility is that we’re using an incorrect kernel version

we may also compare the generator config file offline

i started checking the specific code. 
does the ConfigurableMoE matter for this case?

**Comment by rosong11** [comment]
> does the ConfigurableMoE matter for this case?

The failing case didn’t go through the enable_configurable_moe implementation. The execution parameters were [moe_backend=TRTLLM-mtp_nextn=0-tp4-fp8kv=True-attention_dp=True-cuda_graph=True-overlap_scheduler=True-torch_compile=False].
And I was comparing the outputs of the FlashInfer and TRTLLMGen APIs inside the `run_moe` interface, and their inputs should be the same. 

and how could we compare the generator config file？Thanks.

**Comment by aleozlx** [comment]
i started some wip https://github.com/flashinfer-ai/flashinfer/pull/2425

btw do you use the `trtllm_fp4_block_scale_routed_moe api`?

do you think the packed approach may hit endianess issue? like this https://github.com/flashinfer-ai/flashinfer/pull/2382#issuecomment-3797411121

i'll be back 19:30 pacific time. let's sync up then or later this evening!

**Comment by aleozlx** [comment]
in the wip PR above i started adding the api to allow us the separate param format. it should be supported at the routing runner level. will test it soon.

**Comment by rosong11** [comment]
Hi, @aleozlx What’s the current status of the PR that adds the two parameters? Are we able to pull it in and start testing yet? Thanks.

**Comment by aleozlx** [comment]
hi yes it is ready to be tried

the usage is `trtllm_fp4_block_scale_routed_moe((topk_ids, topk_weights), ...)`

Side note: The old way of packed param will stay unchanged, but for our case of separate topk indices and weights we pass a tuple in its place for compability (topk_ids, topk_weights). This way no breaking changes for existing users, same positional param count in the API.

**Comment by aleozlx** [comment]
~looks like the last commit trying to resolve merge conflict is causing some issue. please use this commit to try for now [473fab8](https://github.com/flashinfer-ai/flashinfer/pull/2425/commits/473fab885a78ce23ff2511531b09514fc4a3bf04)~

resolved

**Comment by rosong11** [comment]
Hi, Thanks for your support! I tested this case locally based on the `[aleozlx:topk_weights](https://github.com/aleozlx/flashinfer/tree/topk_weights)` and there’s no longer a core dump. The test case can pass successfully, but there are still some slight differences: the maximum absolute error is about 0.0078, i tested the api in Qwen3-30B model, and the maximum absolute error is zero. Do you think this difference in numerical precision is normal?

**Comment by wenmingw** [comment]
@aleozlx  could you help share the status of this issue? Thanks.

**Comment by aleozlx** [comment]
thanks for the ping! this is now merged
https://github.com/flashinfer-ai/flashinfer/pull/2425
